spi: stm32: spi_ll_stm32: add MSSI and MIDI support

mustafaabdullahk opened the pull request:
Hi everyone

## Adjustable Signal Flow Timing Support

STM32H7xx series mcu does support that feaute.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/a4eec4d9-42e9-48f7-8918-dd026917ace7)

Several parameters can be used to optionally adjust the timing of the master transaction flow. Signal timings for the master node can be adjusted when needed. This is the case when a slave node needs a longer time to wake up from sleep mode following the setup of the SS active edge or when the slave is not able to handle a data flow that is too fast. Up to 15 additional serial clock signal periods delays can be inserted by MIDI or MSSI parameters. Data frames can be optionally interleaved by SS pulses.

Signed-off-by: Mustafa Abdullah Kus [mustafa.kus@sparsetechnology.com](mailto:mustafa.kus@sparsetechnology.com)

erwango commented on dts/bindings/spi/st,stm32h7-spi.yaml:
```diff
@@ -12,3 +12,18 @@ description: |
+    type: int
+    default: 0
+    description: |
+      Only for STM32H7, (Master Inter-Data Idleness) minimum
```
Restriction `Only for STM32H7,` is misleading. This property is made available on `st,stm32h7-spi` compatible, hence will be available for compatible SoCs: H5/H7/MP1/U5/WBA. This is perfectly fine and sufficient

mustafaabdullahk replied:
good point, i did update document :)

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -618,6 +618,49 @@ static int spi_stm32_configure(const struct device *dev,
 	}
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+	const uint32_t mssi_cycles[] = {
```
These table are an expensive way to configure this. Also, it seems that values encoding looks fairly simple.
What about using directly dt values w/o these tables ?

mustafaabdullahk replied:
i totally aggree with you but following values shifting: 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/72e76736-aeac-49cc-a30d-4f869198b77d)
so i didn't bit shift at the start but update for your suggestion

erwango reviewed (changes requested):
Thanks for this addition!
Looks globally good (Device Tree usage was the tricky part) but I'd like to explore some optimization possibility.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -1160,6 +1165,12 @@ static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_spi_subghz),	\
 		(.use_subghzspi_nss =					\
 			DT_INST_PROP_OR(id, use_subghzspi_nss, false),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.midi_clocks =					\
+			DT_INST_PROP_OR(id, midi_clock, 0),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.mssi_clocks =					\
+			DT_INST_PROP_OR(id, mssi_clock, 0),))\
```
Please mind indentation

```suggestion
	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),		\
		(.midi_clocks =						\
			DT_INST_PROP_OR(id, midi_clock, 0),))		\
	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),		\
		(.mssi_clocks =						\
			DT_INST_PROP_OR(id, mssi_clock, 0),))		\
```

mustafaabdullahk replied:
my tab size value did keep as 4, i did update and solve. :)

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -1160,6 +1165,12 @@ static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 			DT_INST_PROP_OR(id, use_subghzspi_nss, false),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.midi_clocks =					\
+			DT_INST_PROP_OR(id, midi_clock, 0),))\
```
As you provide a default value, prop will automatically be generated with default value (0) if not filled by the user.
Then you don't need to use `DT_INST_PROP_OR` and can just use `DT_INST_PROP(id, midi_clock)`.
Same for the other prop.

Note: It should actually be the same for `use_subghzspi_nss` as boolean properties are false if not present, no need for the `_OR` variant. I'll create an issue to get a clean up done.

mustafaabdullahk replied:
i did update `DT_INST_PROP_OR`,  may i solve use_subghzspi_nss issue with another PR if you want assign to me.

erwango replied:
I've created https://github.com/zephyrproject-rtos/zephyr/issues/67651 to fix this globally. You're welcome to have a stab at it!

erwango reviewed (changes requested):
Thanks for the update.
2 minor fixes that I've just seen and we're good to go.

mustafaabdullahk committed:
dts: bindings: st,stm32h7-spi: add mssi and midi property

add Master Inter-Data Idleness and
Master SS Idleness field. That fields
are integers.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>

mustafaabdullahk committed:
drivers: spi: spi_ll_stm32: add midi and mssi impl

Master Inter-Data Idleness and Master SS Idleness
can configure with STM32 low level spi driver apis.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>
